src/Common/src/TypeSystem/Common/TypeSystemConstaintsHelpers.cs

```diff
@@ -0,0 +1,97 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+
+using System.Diagnostics;
+
+namespace Internal.TypeSystem
+{
+    public static class TypeSystemConstraintsHelpers
+    {
+        private static bool VerifyGenericParamConstraint(Instantiation typeInstantiation, Instantiation methodInstantiation, GenericParameterDesc genericParam, TypeDesc instantiationParam)
+        {
+            // Check class constraint
+            if (genericParam.Constraints.HasFlag(GenericConstraints.ReferenceTypeConstraint) && !instantiationParam.IsGCPointer)
```

Comment:
The type system already has `HasReferenceTypeConstraint`, `HasDefaultConstructorConstraint`, `HasNotNullableValueTypeConstraint`. Why not use them?

`Enum.HasFlag` is a stupid API that does two box operations and a bunch of other things before eventually comparing the two ints. Generally avoid.